tests/scripts/analyze_outcomes.py

```diff
@@ -161,41 +141,57 @@ def analyze_outcomes(results: Results, outcomes, args):
 def read_outcome_file(outcome_file):
     """Parse an outcome file and return an outcome collection.
 
-An outcome collection is a dictionary mapping keys to TestCaseOutcomes objects.
-The keys are the test suite name and the test case description, separated
-by a semicolon.
+An outcome collection is a dictionary presentation of the outcome file:
+```
+outcomes = {
+    "<config>": {
+        "successes": frozenset(["<suite_case>", ... ]),
+        "failures": frozenset(["<suite_case>", ...])
+    }
+    ...
+}
+suite_case = "<suite>;<case>"
+```
 """
     outcomes = {}
     with open(outcome_file, 'r', encoding='utf-8') as input_file:
         for line in input_file:
-            (platform, config, suite, case, result, _cause) = line.split(';')
-            key = ';'.join([suite, case])
-            setup = ';'.join([platform, config])
-            if key not in outcomes:
-                outcomes[key] = TestCaseOutcomes()
+            (_platform, config, suite, case, result, _cause) = line.split(';')
```

Comment:
> `config` is not unique

Yes, but currently I think currently we don't have a task that cares about `platform`. That's why I don't use `platform` field at present.

> How does this affect the rest of the script?

`analyze_coverage` only cares about the presents of test cases, it doesn't care if a test case is in `successes` or `failures` or both.
`analyze_driver_vs_reference` could be affected if the components are running on multiple platform, but this is not true for current CI?

Reply:
I agree that currently we don't have any tasks that care about `platform`, or for which it would be a problem to have a case that's both in `successes` and `failures`.

However, before reading Gilles's comment, I had not realised that `config` is not unique and that this could result in a test case being in both `successes` and `failures` set, so I agree that this deserves at least a comment.

Reply:
Btw, not directly related, but this made me think: I think what both functions care about is not really the distinction between "success" and "failure", but the distinction between "executed" and "skipped": for `analyze_coverage` that's pretty clear, and for `analyze_driver_vs_reference` even though the docstring currently says "passing" what this is really about is that we don't want tests being skipped in the driver component unless we know it's expected and justified.

Generally speaking, failures will be found and reported at earlier stages in the CI, so I think this script is really about detecting unexpected skips, not about failures.

Reply:
> what both functions care about is not really the distinction between "success" and "failure", but the distinction between "executed" and "skipped"

I agree. But I consider this optional here.

When I wrote the original script, I recorded the pass/fail information because it was very easy and I didn't know how the script would evolve beyond the initial feature of test case coverage. After a few years, it turns out we don't need the pass/fail information, but it doesn't really harm to keep it (it isn't really increasing the complexity).